flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialects.java

```diff
@@ -18,23 +18,25 @@
 
 package org.apache.flink.connector.jdbc.dialect;
 
-import java.util.Arrays;
-import java.util.List;
+import org.apache.flink.annotation.Internal;
+
 import java.util.Optional;
+import java.util.ServiceLoader;
 
 /** Default JDBC dialects. */
+@Internal
 public final class JdbcDialects {
 
-    private static final List<JdbcDialect> DIALECTS =
-            Arrays.asList(new DerbyDialect(), new MySQLDialect(), new PostgresDialect());
-
     /** Fetch the JdbcDialect class corresponding to a given database url. */
     public static Optional<JdbcDialect> get(String url) {
-        for (JdbcDialect dialect : DIALECTS) {
-            if (dialect.canHandle(url)) {
-                return Optional.of(dialect);
+        ClassLoader cl = Thread.currentThread().getContextClassLoader();
+
+        for (JdbcDialectFactory factory : ServiceLoader.load(JdbcDialectFactory.class, cl)) {
+            if (factory.canHandle(url)) {
```

Comment:
What if there are multiple implementations that accept the URL? We should probably catch such cases and give an appropriate error. 

Comparing with `FactoryUtil` for the Table & SQL stack, it probably also makes sense to throw an error if no implementations could be found (note: I don't mean none that accept the URL, but none at all, hinting to some setup problem).

Reply:
yep, I'm just going to copy the logic of FactoryUtil. I also did a pass and realized nothing that uses this class actually handles `Optional.empty` gracefully, they all just throw. That makes it simpler to just throw if we don't find exactly one dialect on the classpath.